### PR TITLE
Improve ConfusionMatrix for binary classification

### LINKED ISSUE #1380: Documentation for ConfusionMatrix is thin on details, and error messages not informative
## 🐛 Bug description

Confusion matrix requires that `y_pred` and `y` are to have different dimensions `(batch,num_classes)` and `(batch,)`, respectively. Getting this wrong results in the following error message, which makes no sense (batch size is 5 here, and both `y_pred` and `y` have shape `(5,1)`:

`y_pred must have shape (batch_size, num_categories, ...) and y must have shape of (batch_size, ...), but given torch.Size([5, 1]) vs torch.Size([5, 1])`

I believe that the `torch.Size([5, 1]) vs torch.Size([5, 1])` message should probably read `torch.Size([5, 1]) vs torch.Size([5,])` to illustrate the problem.

Separately, it would be good to mention in the `ConfusionMatrix` docs that one-hot encoded vectors are required, and that this can be implemented using `from ignite.utils.to_onehot` (I stumbled upon it after realizing I had to one-hot encode, since most people use `BCEWithLogitsLoss()` anyway, and hence their model predictions are not one-hot encoded).

## Environment

 - PyTorch Version (e.g., 1.4): 1.4.0
 - Ignite Version (e.g., 0.3.0): 0.4.2
 - OS (e.g., Linux): MacOS
 - How you installed Ignite (`conda`, `pip`, source): `pip`
 - Python version: 3.8
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by jrieke**
Fixes #1380 

Description:

Makes confusion matrix work better with binary classification. Adds a check that the `num_classes` arg of `ConfusionMatrix` is >1 (and tests for that), improves other error messages, and adds an example for how to transform the network output for binary classification. See [this comment](https://github.com/pytorch/ignite/issues/1380#issuecomment-768681135) for more details.

Check list:
* [x] New tests are added (if a new feature is added)
* [x] New doc strings: description and/or example code are in RST format
* [ ] Documentation is updated (if required)

### CODE REVIEW

File: ignite/metrics/confusion_matrix.py [hunk 1]
```diff
@@ -14,13 +14,14 @@ class ConfusionMatrix(Metric):
...
-    - `y_pred` must contain logits and has the following shape (batch_size, num_categories, ...)
+    - `y_pred` must contain logits and has the following shape (batch_size, num_classes, ...). 
+      If you are doing binary classification, see Note for an example on how to get this.
     - `y` should have the following shape (batch_size, ...) and contains ground-truth class indices
       with or without the background class. During the computation, argmax of `y_pred` is taken to determine
       predicted classes.
 
     Args:
-        num_classes (int): number of classes. See notes for more details.
+        num_classes (int): Number of classes, should be >1. See notes for more details.
```
Comment (vfdev-5) [inline ignite/metrics/confusion_matrix.py:24]:
```suggestion
        num_classes (int): Number of classes, should be > 1. See notes for more details.
```

File: ignite/metrics/confusion_matrix.py [hunk 2]
```diff
@@ -50,6 +68,9 @@ def __init__(
     ):
         if average is not None and average not in ("samples", "recall", "precision"):
             raise ValueError("Argument average can None or one of 'samples', 'recall', 'precision'")
+        
+        if num_classes <= 1:
+            raise ValueError("Argument num_classes needs to be >1")
```
Comment (vfdev-5) [inline ignite/metrics/confusion_matrix.py:73]:
```suggestion
            raise ValueError("Argument num_classes needs to be > 1")
```

File: tests/ignite/metrics/test_confusion_matrix.py [hunk 3]
```diff
@@ -19,17 +19,22 @@ def test_no_update():
         cm.compute()
 
 
+def test_num_classes_wrong_input():
+    with pytest.raises(ValueError, match="Argument num_classes needs to be >1"):
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_confusion_matrix.py:23]:
```suggestion
    with pytest.raises(ValueError, match="Argument num_classes needs to be > 1"):
```

**Review by vfdev-5 — COMMENTED** [review]
Looks good! Remains to fix code-formatting and we are good to go.

See https://github.com/pytorch/ignite/blob/master/CONTRIBUTING.md#formatting-without-pre-commit

File: ignite/metrics/confusion_matrix.py [hunk 4]
```diff
@@ -38,6 +39,22 @@ class ConfusionMatrix(Metric):
         In case of the targets `y` in `(batch_size, ...)` format, target indices between 0 and `num_classes` only
         contribute to the confusion matrix and others are neglected. For example, if `num_classes=20` and target index
         equal 255 is encountered, then it is filtered out.
+        
+        If you are doing binary classification with a single output unit, you may have to transform your network output,
+        so that you have one value for each class. E.g. you can transform your network output into a one-hot vector with:
+
+            def binary_one_hot_output_transform(output):
```
[the remark this answers is not in the document]
Comment (jrieke) [reply ignite/metrics/confusion_matrix.py:49]: Done

**Comment by nsinenian** [comment]
This looks great to me, thanks everyone!

**Pull request merged**